### DISCUSSION TITLE
classes/class_rigidbody2d

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#43] [User-contributed notes]
# classes/class_rigidbody2d

Inherits: PhysicsBody2D< CollisionObject2D< Node2D< CanvasItem< Node< Object Inherited By: PhysicalBone2D A 2D physics body that is moved by a physics simulation. Description: RigidBody2D implement...

https://docs.godotengine.org/en/stable/classes/class_rigidbody2d.html

**Comment by ottworks** [comment]
Careful with apply_force - it expects neither local coordinates nor global coordinates. it expects the *offset* in global coordinates, e.g. the global coordinates minus the rigidbody's  global_position.

**Comment by psychicteeth** [comment]
> Adds a constant positioned force to the body that keeps being applied over time until cleared with constant_force = Vector2(0, 0).

~~This is not correct. Setting `constant_force` to zero does not stop the body from moving, at least in _process.~~

**Reply by Calinou** [reply]
The constant positioned force is cleared, not the body's velocity. In other words, it's relative. You can see it as a "boost" in one direction that is no longer applied when set to `(0, 0)` (since `x + 0 = x`).

Internally, this is what GodotPhysics does when calling `add_constant_force()`/`add_constant_central_force()`:

https://github.com/godotengine/godot/blob/af77100e394dcaca609b15bef815ed17475e51ed/modules/godot_physics_3d/godot_body_3d.h#L264-L271

**Reply by psychicteeth** [reply]
Thanks. I think what I was experiencing the other day was an acceleration due to gravity actually, which appeared to be a constant force due to friction. Silly me!

**Comment by follyfoxe** [comment]
If you're struggling to teleport a Rigidbody2D, the following worked for me:
```gdscript
func teleport(rb : RigidBody2D, target : Transform2D) -> void:
	rb.linear_velocity = Vector2.ZERO
	rb.angular_velocity = 0.0
	rb.global_transform = target
	PhysicsServer2D.body_set_state(rb.get_rid(), PhysicsServer2D.BODY_STATE_TRANSFORM, target)
	rb.reset_physics_interpolation()
```
Any other way I tried the body kept colliding with objects, as if the body were moving to the destination instead of teleporting.

**Reply by HenriSup** [reply]
You saved me hours of research and time trying to understand all this ! 
If people still experience strange behavior, remember that you might need to also reset childrens' Rigidbodies Angular velocity and linear velocity (I had to do that to the wheels of my car so it doesn't keep momentum when teleporting).

Thanks again you're a Hero :D

**Reply by BinaryGleam** [reply]
Wouldn't it also be possible to use `_integrate_forces(state)` function to do so? As the description of the function says:
"*Called during physics processing, allowing you to read and safely modify the simulation state for the object. By default, it is called before the standard force integration, but the custom_integrator property allows you to disable the standard force integration and do fully custom force integration for a body.*"

Also in [using rigibodies](https://docs.godotengine.org/en/stable/tutorials/physics/rigid_body.html#using-rigidbody), the function is described as follow:
"*The fact that you can't use set_global_transform() or look_at() methods doesn't mean that you can't have full control of a rigid body. Instead, you can control it by using the _integrate_forces() callback. In this method, you can add forces, apply impulses, or set the velocity in order to achieve any movement you desire.*"

So I think that would be the key to teleport a RB in a way that is coherent with godot's documentation? But I haven't tried it so maybe im spitballing nothingness here.

**Reply by follyfoxe** [reply]
@BinaryGleam most likely, yeah. You could theoretically use the integrate forces function, but the problem is you don't always have access to it! So this solution is perfect for when you can't attach scripts to your bodies. It is also, instant, whereas with integrate forces you'd have to wait until that function is called. But it'd be a totally valid way to approach it, I believe.

**Reply by BinaryGleam** [reply]
Very insightful answer that shows the strength of your method! Thank you for taking the time to explain it even further, it is appreciated!

**Comment by BinaryGleam** [comment]
In the documentation for contact monitoring, it says under the contact_monitor variable this:
```gdscript
If true, the RigidBody2D will emit signals when it collides with another body.
```
Which signal? Like how do I catch the relevant informations when a contact happens?

**Reply by BinaryGleam** [reply]
Turns out it's the body_enter signal, I was in a specific case where going from a frozen to an unfrozen rigidbody switched monitoring on and off. I feel dumb now

**Comment by nuke-haus** [comment]
Be wary of using SetGlobalPosition with a RigidBody2D. In order for the rigid body to not snap back to where it previously was, you need to grab its GlobalTransform, update the Origin of this transform, then use SetGlobalTransform with that value.
